src/main/java/picard/sam/markduplicates/UmiAwareDuplicateSetIterator.java

```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Override
+    public DuplicateSet next() {
+        if(!nextSetsIterator.hasNext())
+            process(wrappedIterator.next());
+
+        return nextSetsIterator.next();
+    }
+
+    // Takes a duplicate set and breaks it up into possible smaller sets according to the UMI,
+    // and updates nextSetsIterator to be an iterator on that set of DuplicateSets.
+    private void process(final DuplicateSet set) {
+
+        final List<SAMRecord> records = set.getRecords();
+
+        // If any records are missing the UMI_TAG proceed as if there were no UMIs
+        // and return nextSetsIterator without breaking it up into smaller sets.
+        for(SAMRecord rec : records) {
+            if(rec.getStringAttribute(umiTag) == null) {
+                nextSetsIterator = Collections.singleton(set).iterator();
+                return;
+            }
+        }
+
+        // Count all the uniquely observed UMI sequences
+        Map<String, Long> umiFrequencies = records.stream()
+                        .collect(Collectors.groupingBy(p -> p.getStringAttribute(umiTag),
+                                counting()));
+
+        UmiGraph umiGraph = new UmiGraph();
+
+        // Add each of the uniquely observed UMI sequences to the graph
+        int firstGroup = 0;
+        for (Map.Entry<String, Long> umiFrequency : umiFrequencies.entrySet()) {
+            umiGraph.addNode(new UmiSequence(umiFrequency.getKey(), umiFrequency.getValue()));
+        }
+        umiGraph.processNodes();
+
+        Map<Integer, List<UmiSequence>> umisByDuplicateSet = umiGraph.getUmisByDuplicateSet();
+
+        // Construct DuplicateSetList
+        List<DuplicateSet> duplicateSetList = new ArrayList<>();
+        for(int i = 0;i < umiGraph.duplicateSetCounter;i++) {
+            DuplicateSet e = new DuplicateSet();
+            duplicateSetList.add(e);
+        }
+
+        for(SAMRecord rec : records) {
+            String umi = records.get(records.indexOf(rec)).getStringAttribute(umiTag);
+
+            // Figure out which group this belongs to
+            int recordGroup = umiGraph.node.get(umi).duplicateSet;
+            duplicateSetList.get(recordGroup).add(records.get(records.indexOf(rec)));
+        }
+
+        // Optionally add the inferred Umi
+        if(addInferredUmi) {
+            for(DuplicateSet ds : duplicateSetList) {
+                // For each duplicate set identify the most common UMI
+                List<SAMRecord> recordsFromDuplicateSet = ds.getRecords();
+
+                // Assign the inferred UMI to all reads in the current group
+                for(SAMRecord record : recordsFromDuplicateSet) {
+                    record.setAttribute(inferredUmiTag, umiGraph.getInferredUmiFromOriginalSequence(record.getStringAttribute(umiTag)));
+                }
+            }
+        }
+
+        nextSetsIterator = duplicateSetList.iterator();
+    }
+
+    private int getEditDistance(final String s1, final String s2) {
```

Comment:
there is atleast one other place that calculates edit distance...could you refactor this to a stringUtil method?